profiler/internal/fastdelta/fd.go

```diff
@@ -0,0 +1,824 @@
+	switch ProfileRecordNumber(field) {
+	case recProfileSampleType:
+		vType, vUnit, err := dc.readValueType(value.Bytes)
+		if err != nil {
+			return false, fmt.Errorf("reading sample_type record: %w", err)
+		}
+		dc.valueTypeIndices = append(dc.valueTypeIndices, [2]int{vType, vUnit})
+	case recProfileLocation:
+		// readLocation writes out function IDs for the location to this scratch buffer
+		dc.scratchIDs = dc.scratchIDs[:0]
+		address, id, mappingID, err := dc.readLocation(value.Bytes)
+		if err != nil {
+			return false, fmt.Errorf("reading location record: %w", err)
+		}
+		dc.locationIndex.Insert(id, address, mappingID, dc.scratchIDs)
+	case recProfileStringTable:
+		dc.strings.add(value.Bytes)
+
+		// always include the zero-index empty string,
+		// otherwise exclude by default unless used by a kept sample in mergeSamplesPass
+		dc.includeString = append(dc.includeString, len(dc.includeString) == 0)
+	}
+	return true, nil
+}
+
+// mergeSamplesPassFn returns a molecule callback to scan a Profile protobuf
+// and write merged samples to the output buffer.
+// Any samples where the values are all 0 will be skipped.
+// This pass has the side effect of populating dc.include* fields so only the
+// mapping, function, location, and strings (sample labels) referenced by the kept samples
+// can be written out in a later pass.
+func (dc *DeltaComputer) mergeSamplesPassFn(out io.Writer, hasher murmur3.Hash128) molecule.MessageEachFn {
+	var sampleHash Hash
+
+	computeDeltaForValue := make([]bool, len(dc.valueTypeIndices))
+	for _, field := range dc.fields {
+		for i, vtIdxs := range dc.valueTypeIndices {
+			typeMatch := dc.strings.Equals(vtIdxs[0], []byte(field.Type))
+			unitMatch := dc.strings.Equals(vtIdxs[1], []byte(field.Unit))
+			if typeMatch && unitMatch {
+				computeDeltaForValue[i] = true
+				break
+			}
+		}
+	}
+
+	return func(field int32, value molecule.Value) (bool, error) {
+		return dc.mergeSamplesPass(field, value, out, hasher, &sampleHash, computeDeltaForValue)
+	}
+}
+
+func (dc *DeltaComputer) mergeSamplesPass(
+	field int32,
+	value molecule.Value,
+	out io.Writer,
+	hasher murmur3.Hash128,
+	sampleHash *Hash,
+	computeDeltaForValue []bool) (bool, error) {
+	if ProfileRecordNumber(field) != recProfileSample {
+		return true, nil
+	}
+
+	// readSample writes out the locations for this sample into this scratch buffer
+	// to save on allocations
+	dc.scratchIDs = dc.scratchIDs[:0]
+	val, err := dc.readSample(value.Bytes, hasher, sampleHash)
+	if err != nil {
+		return false, fmt.Errorf("reading sample record: %w", err)
+	}
+
+	old, ok := dc.sampleMap[*sampleHash]
+	dc.sampleMap[*sampleHash] = val // save for next time
+	if !ok {
+		// If this is a new sample we don't take the
+		// difference, just pass it through.
+		// but we should record the value for next time
+		if err := dc.writeProtoBytes(out, field, value.Bytes); err != nil {
+			return false, err
+		}
+		dc.keepLocations(dc.scratchIDs)
+		return true, nil
```

Comment:
🐞 Since there is no `includeStringIndexFields` call here, samples with labels that go through this code pass may have string table references that don't exist.

Reply:
good catch, will fix

Reply:
> good catch
> In practice, we throw away the first delta and the comparingDeltaProfiler didn't catch this one, but will fix.

Maybe hold off on that fix. I have a branch with some more refactorings where I can easily fix this as well.